pkgs/build-support/rust/default.nix

```diff
@@ -66,14 +72,126 @@ let
     else ''
       cargoDepsCopy="$sourceRoot/${cargoVendorDir}"
     '';
+  
+  targetIsJSON = stdenv.lib.hasSuffix ".json" target;
+
+  rustSrc = stdenv.mkDerivation {
+    name = "rust-src";
+    src = rustPlatform.rust.rustc.src;
+    preferLocalBuild = true;
+    phases = [ "unpackPhase" "installPhase" ];
+    installPhase = "cp -r src $out";
+  };
+
+  toTOML = value: builtins.readFile (runCommandNoCC "tmp.toml" {
+    nativeBuildInputs = [ remarshal ];
+    value = builtins.toJSON value;
+    passAsFile = [ "value" ];
+  } ''
+    json2toml "$valuePath" "$out"
+  '');
+
+  # sysroot logic based on https://github.com/DianaNites/cargo-sysroot/
+
+  sysrootSrc = let
+    parentToml = builtins.readFile (src + /Cargo.toml);
+    parentTomlParsed = builtins.fromTOML parentToml;
+    parentProfile = if builtins.hasAttr "profile" parentTomlParsed
+                    then parentTomlParsed.profile else {};
+    sysrootProfileToml = toTOML parentProfile;
+  in stdenv.mkDerivation {
```

Comment:
I would prefer if this parsing logic would be moved to the build time rather than evaluation time. Otherwise whenever nix wants to build this package it would also need to download the source for it, which is expensive on hydra and the final client that tries to download the binary package from the cache. Within nixpkgs we don't allow import-from-derivation for this reason.